py/picca/pk1d/postproc_pk1d.py

```diff
@@ -862,12 +818,370 @@ def compute_average_pk_redshift(
+
+    func = partial(
+        compute_groups,
+        nbins_k,
+    )
+    if number_worker == 1:
+        output_cov = [func(*p1d_los) for p1d_los in p1d_los_table]
+    else:
+        with Pool(number_worker) as pool:
+            output_cov = np.array(pool.map(func, p1d_los_table))
+
+    del p1d_los_table
+
+    p1d_weights, covariance_weights, p1d_groups = (
+        output_cov[:, 0, :],
+        output_cov[:, 1, :],
+        output_cov[:, 2, :],
+    )
+    return p1d_weights, covariance_weights, p1d_groups
+
+
+def compute_groups(nbins_k, p1d_los):
+    """Compute the P1D groups for one subforest.
+
+    Arguments
+    ---------
+    nbins_k (int):
+    Number of k bins.
+
+    p1d_los (array-like):
+    Table containing all p1d pixels unordered for one subforest
+
+    Return
+    ------
+    p1d_weights_id  (array-like):
+    Weights associated with p1d pixels for one subforest, used in the calculation of covariance.
+
+    covariance_weights_id (array-like):
+    Weights for one subforest used inside the main covariance sum.
+
+    p1d_groups_id (array-like):
+    Individual p1d pixels grouped in the same wavenumber binning for one subforest
+    """
+    p1d_weights_id = np.zeros(nbins_k)
+    covariance_weights_id = np.zeros(nbins_k)
+    p1d_groups_id = np.zeros(nbins_k)
+
+    for ikbin in range(nbins_k):
+        mask_ikbin = p1d_los["k_index"] == ikbin
+        number_in_bins = len(mask_ikbin[mask_ikbin])
+        if number_in_bins != 0:
+            weight = p1d_los["weight"][mask_ikbin][0]
+            p1d_weights_id[ikbin] = weight
+            covariance_weights_id[ikbin] = np.sqrt(weight / number_in_bins)
+            p1d_groups_id[ikbin] = np.nansum(
+                p1d_los["pk"][mask_ikbin] * covariance_weights_id[ikbin]
+            )
+    return p1d_weights_id, covariance_weights_id, p1d_groups_id
+
+
 def compute_cov(
+    weight_method,
+    mean_p1d_table,
+    nbins_k,
+    izbin,
+    p1d_weights,
+    covariance_weights,
+    p1d_groups,
+):
+    """Compute the covariance of a set of 1D power spectra.
+
+    Arguments
+    ---------
+    weight_method: str,
+    Method to weight the data.
+
+    mean_p1d_table (array-like):
+    Table of mean 1D power spectra, with column 'meanPk'.
+
+    nbins_k (int):
+    Number of k bins.
+
+    izbin (int):
+    Current redshift bin being considered.
+
+    p1d_weights  (array-like):
+    Weights associated with p1d pixels for all subforest, used in the calculation of covariance.
+
+    covariance_weights (array-like):
+    Weights for all subforest used inside the main covariance sum.
+
+    p1d_groups (array-like):
+    Individual p1d pixels grouped in the same wavenumber binning for all subforest
+
+    Return
+    ------
+    covariance_array (array-like):
+    Array of covariance coefficients.
+    """
+
+    if len(p1d_groups) == 0:
+        return np.full(nbins_k * nbins_k, np.nan)
+
+    mean_pk = mean_p1d_table["meanPk"][(nbins_k * izbin) : (nbins_k * (izbin + 1))]
```

Comment:
given that one only really needs the relevant slice for the redshift bin in this function, why not pass that slice instead of izbin? That would avoid needing to define the slicing operation several times throughout this code...
Alternatively having functions computing the p1d_table_slice and the cov_table_slice would make the code more flexible later on (when we might deal with different number of k-bins per redshift), e.g.:
```
def p1d_table_slice(izbin, nbins=nbins_k):
     return slice((nbins * izbin), (nbins * (izbin + 1))
```

Reply:
But you are right concerning the first comment, I will pass the two slices

Reply:
oh, with the function definition I just wanted to introduce something that would work as the current implementation, but that could be changed at a single place easily later if the approach should be changed...

Reply:
Ah yes this is indeed a good way, it is done now.